components/normalizer/src/lib.rs

```diff
@@ -2529,97 +2546,109 @@ impl<'data> ComposingNormalizerBorrowed<'data> {
+                                    if in_inclusive_range16(low, 0xDC00, 0xDFFF) {
+                                        upcoming32 = (upcoming32 << 10) + u32::from(low)
+                                            - (((0xD800u32 << 10) - 0x10000u32) + 0xDC00u32);
+                                        // Successfully-paired surrogate. Read from the trie again.
+                                        trie_value = composition.decomposition.trie.get32(upcoming32);
+                                        if potential_passthrough_and_cannot_combine_backwards_impl(trie_value) {
+                                            // Fast-track succeeded!
+                                            continue 'fast;
+                                        }
+                                        break 'surrogateloop;
+                                    } else {
+                                        code_unit_iter = iter_backup;
                                     }
-                                    break 'surrogateloop;
-                                } else {
-                                    code_unit_iter = iter_backup;
                                 }
                             }
+                            // unpaired surrogate
+                            upcoming32 = 0xFFFD; // Safe value for `char::from_u32_unchecked` and matches later potential error check.
+                            // trie_value already holds a decomposition to U+FFFD.
+                            debug_assert_eq!(trie_value, NON_ROUND_TRIP_MARKER | BACKWARD_COMBINING_MARKER | 0xFFFD);
+                            break 'surrogateloop;
                         }
-                        // unpaired surrogate
-                        upcoming32 = 0xFFFD; // Safe value for `char::from_u32_unchecked` and matches later potential error check.
-                        // trie_value already holds a decomposition to U+FFFD.
-                        debug_assert_eq!(trie_value, NON_ROUND_TRIP_MARKER | BACKWARD_COMBINING_MARKER | 0xFFFD);
-                        break 'surrogateloop;
-                    }
 
-                    // SAFETY: upcoming32 can no longer be a surrogate.
-                    let upcoming = unsafe { char::from_u32_unchecked(upcoming32) };
-                    let upcoming_with_trie_value = CharacterAndTrieValue::new(upcoming, trie_value);
-                    // We need to fall off the fast path.
-                    composition.decomposition.pending = Some(upcoming_with_trie_value);
-                    #[expect(clippy::indexing_slicing)]
-                    let mut consumed_so_far = pending_slice[..pending_slice.len() - code_unit_iter.as_slice().len() - upcoming.len_utf16()].chars();
-                    // `unwrap` OK, because we've previously managed to read the previous character
-                    #[expect(clippy::unwrap_used)]
-                    {
+                        // SAFETY: upcoming32 can no longer be a surrogate.
```

Comment:
nit: please mark all the other comments above around surrogateness with SAFETY, and justify it at every loop exit

Reply:
Isn't this out of scope for this changeset, since this in pre-existing code that merely gets re-indended?